src/ZEO/asyncio/tests.py

```diff
@@ -917,17 +1246,555 @@ def it(tag):
+        self.assertTrue(t.done())
+        self.assertEqual(t.result(), 123)
+
+        # nested returns
+        @coroutine
+        def g():
+            x = yield f()
+            return_(1000 + x)
+            yield
+
+        t = self.make_task(g)
+        self.assertTrue(t.done())
+        self.assertEqual(t.result(), 1123)
+
+        # verify that return can go through `except Exception`
+        # (but not through `except BaseException`)
+        @coroutine
+        def h():
+            try:
+                return_(456)
+                raise Exception('aaa')
+            except Exception:
+                raise
+            yield
+
+        t = self.make_task(h)
+        self.assertTrue(t.done())
+        self.assertEqual(t.result(), 456)
+
+
+class AsyncTaskTests(CoroutineExecutorTestsBase, TestCase):
+    def make_task(self, coro):
+        return AsyncTask(coro(), self.loop)
+
+
+class ConcurrentTaskTests(CoroutineExecutorTestsBase, TestCase):
+    def make_task(self, coro):
+        loop = self.loop
+        t = ConcurrentTask(coro(), loop)
+        self.run_loop()
+        return t
+
+    def test_cancel_from_another_thread(self):
+        l = []
+        waitready = threading.Event()
+        waiting = Future(loop=self.loop)
+
+        @coroutine
+        def f():
+            l.append(1)
+            l.append(2)
+            waitready.set()
+            yield waiting
+            l.append(3)
+
+        t = self.make_task(f)
+        self.assertFalse(t.done())
+
+        def Tcancel():
+            waitready.wait()
+            t.cancel('zzz')
+        tcancel = threading.Thread(target=Tcancel)
+        tcancel.setDaemon(True)
+        tcancel.start()
+
+        with self.assertRaises(asyncio.CancelledError):
+            loop_run_until_complete(self.loop, t)
+        tcancel.join(9)
+        self.assertFalse(tcancel.is_alive())
+        self.assertTrue(t.done())
+        self.assertTrue(t.cancelled())
+        with self.assertRaises(asyncio.CancelledError) as e:
+            t.result()
+        self.assertEqual(e.exception.args, ('zzz',))
+        self.assertTrue(waiting.done())
+        self.assertTrue(waiting.cancelled())
+        self.assertEqual(l, [1,2])
+        _ = t.cancel()
+        self.assertFalse(_)
+
 
 def test_suite():
     suite = unittest.TestSuite()
     suite.addTest(unittest.makeSuite(ClientTests))
     suite.addTest(unittest.makeSuite(ServerTests))
     suite.addTest(unittest.makeSuite(MsgpackClientTests))
     suite.addTest(unittest.makeSuite(MsgpackServerTests))
-    suite.addTest(unittest.makeSuite(ProtocolTests))
+    suite.addTest(unittest.makeSuite(ZEOBaseProtocolTests))
+    suite.addTest(unittest.makeSuite(SizedMessageProtocolTests))
+    suite.addTest(unittest.makeSuite(FutureTests))
+    suite.addTest(unittest.makeSuite(ConcurrentFutureTests))
+    suite.addTest(unittest.makeSuite(AsyncTaskTests))
+    suite.addTest(unittest.makeSuite(ConcurrentTaskTests))
     return suite
```

Comment:
`unittest.makesuite` is deprecated. Likely, we can drop the whole definition of `test_suite`.

Reply:
Yes, that's what I meant with 

---- 8< ----
b) removal of `def test_suite()` is `src/ZEO/asyncio/test.py`.

   Those `test_suite()` functions are there to support `python setup.py test`.
   I agree this way of running tests is not used nowdays, but if we are to drop
   it, we should do it similarly to
   https://github.com/zopefoundation/ZODB/commit/3577a9c4 - i.e. dropping
   support for `python setup.py test` completely instead of making it broken.

   I can do corresponding PR if we decide that dropping `python setup.py test` is ok.
---- 8< ----

in https://github.com/zopefoundation/ZEO/pull/195#issuecomment-1374901652 .